hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java

```
@@ -451,7 +451,50 @@ private File downloadSignedDataFile(File sigFile, String nodeAccountId) {
         return null;
     }
 
-    protected abstract boolean verifyDataFile(File file, byte[] signedHash);
+    private StreamFile readAndVerifyStreamFile(File file, String verifiedHash) {
+        String fileName = file.getName();
+        StreamFile streamFile = readStreamFile(file);
+
+        if (lastValidDownloadedFileHashKey != null) {
+            String expectedPrevFileHash = applicationStatusRepository.findByStatusCode(lastValidDownloadedFileHashKey);
+            Instant verifyHashAfter = downloaderProperties.getMirrorProperties().getVerifyHashAfter();
+            if (!verifyHashChain(streamFile.getPreviousHash(), expectedPrevFileHash, verifyHashAfter, fileName)) {
+                throw new HashMismatchException(fileName, expectedPrevFileHash, streamFile.getPreviousHash());
+            }
+        }
+
+        if (!streamFile.getFileHash().contentEquals(verifiedHash)) {
+            throw new HashMismatchException(fileName, verifiedHash, streamFile.getFileHash());
+        }
+
+        return streamFile;
+    }
+
+    /**
+     * Updates last valid downloaded file and last valid downloaded file hash key in database if applicable. Also saves
+     * the stream file to its corresponding database table.
+     * @param streamFile the verified stream file
+     */
+    private void updateApplicationStatus(StreamFile streamFile) {
+        transactionTemplate.executeWithoutResult(status -> {
+            try {
+                if (lastValidDownloadedFileHashKey != null) {
+                    applicationStatusRepository
+                            .updateStatusValue(lastValidDownloadedFileHashKey, streamFile.getFileHash());
+                }
+                applicationStatusRepository.updateStatusValue(lastValidDownloadedFileKey, streamFile.getName());
+
+                saveStreamFileRecord(streamFile);
+            } catch (Exception ex) {
+                log.error("Roll back the transaction due to error", ex);
```

This is a change in behavior as it won't try another node now. We shouldn't catch and rollback here manually. Let TransactionTemplate automatically rollback on exception and let the exception bubble up so it can retry the next node.

---

The catch and log here is intended to log the original exception so we have the complete stack trace. I'll run it manually to see if it's not needed and the TransactionException thrown by TransactionTemplate has all info.

The logic here does not change the behavior if one node fails we continue with the next in the list because the outer try... with... is inside the for loop.

---

removed the catch and log